routing.ts

```diff
@@ -777,12 +785,44 @@ namespace jacdac {
 
         protected readonly config: ClientPacketQueue
         private readonly registers: RegisterClient<PackDataType[]>[] = []
+        readonly _role: string
 
-        constructor(public readonly serviceClass: number, public role: string) {
+        constructor(public readonly serviceClass: number, role: string) {
             super()
+            if (!role) throw "no role"
+
             this.eventId = control.allocateNotifyEvent()
             this.config = new ClientPacketQueue(this)
-            if (!this.role) throw "no role"
+            this._role = role
+        }
+
+        get role() {
+            const i = this._role.indexOf("?")
+            if (i > -1) return this._role.substr(0, i)
+            else return this._role
+        }
+
+        get roleQuery(): ClientRoleQuery {
+            const i = this._role.indexOf("?")
+            if (i < 0) return undefined
```

Comment:
you need to return new ClientRoleQuery(this._role) here, otherwise the call to matchesRoleAt below will fail